geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionImpl.java

```diff
@@ -387,14 +387,20 @@ private void checkCancelled() {
         Throwable th = it_causes.next();
 
         if (!membership.hasMember(member) || (th instanceof ShunnedMemberException)) {
+          if (logger.isDebugEnabled()) {
+            logger.debug(String.format("Failed to send message <%s> to member <%s> view, %s",
+                content, member, membership.getView()), th);
```

Comment:
I can see some value in this message, but I have some suggestions.

First, think about the message content. As written, it doesn't tell why it failed. Why not be explicit with something like: "Failed to send message to member <%s> no longer in view."

Second, think about the frequency of this message. I think it'll happen whenever members are leaving (shutting down, crashing, cycling). Not all that frequent so I guess that's ok. OTOH message volume can be very high. So the product of those can be high.

Given we are adding potentially lots of logging, I don't think we need the view in this message. The person reading the logs can search backward for the previous view.

And do we really need the content in this message? I think that can be fairly large too.

Reply:
The test changes in my opinion don't make the failure easier to understand or diagnose.  They do improve the overall test quality, ensure we are testing the right thing, and reduce flakiness to some degree.  It would be nice if there was a test that more directly tested the GEODE-9350 issue. 

If this issue or something similar happened again (in this test or somewhere else), it would be nice to have a debug log that might help us figure it out.  I've updated it to be more concise, but it does still log the exception.  We could potentially remove that or just get the message from it if it seems like too much:
```
logger.debug(String.format("Failed to send message <%s> to member <%s> no longer in view",
                content.getShortClassName(), member), th);
```

In the logs, this looks like:
```
Failed to send message <MemberFunctionStreamingMessage> to member <26461460c57d(172)<v2>:41002> no longer in view
org.apache.geode.distributed.internal.direct.ShunnedMemberException: Member is being shunned: 26461460c57d(172)<v2>:41002
        at org.apache.geode.distributed.internal.direct.DirectChannel.getConnections(DirectChannel.java:467)
        at org.apache.geode.distributed.internal.direct.DirectChannel.sendToMany(DirectChannel.java:281)
        at org.apache.geode.distributed.internal.direct.DirectChannel.sendToOne(DirectChannel.java:188)
        at org.apache.geode.distributed.internal.direct.DirectChannel.send(DirectChannel.java:537)
        at org.apache.geode.distributed.internal.DistributionImpl.directChannelSend(DistributionImpl.java:354)
        at org.apache.geode.distributed.internal.DistributionImpl.send(DistributionImpl.java:296)
        at org.apache.geode.distributed.internal.ClusterDistributionManager.sendViaMembershipManager(ClusterDistributionManager.java:2067)
        at org.apache.geode.distributed.internal.ClusterDistributionManager.sendOutgoing(ClusterDistributionManager.java:1982)
        at org.apache.geode.distributed.internal.ClusterDistributionManager.sendMessage(ClusterDistributionManager.java:2027)
        at org.apache.geode.distributed.internal.ClusterDistributionManager.putOutgoing(ClusterDistributionManager.java:1084)
        at org.apache.geode.internal.cache.execute.StreamingFunctionOperation.getFunctionResultFrom(StreamingFunctionOperation.java:113)
        at org.apache.geode.internal.cache.execute.MemberFunctionExecutor.executeFunction(MemberFunctionExecutor.java:149)
        at org.apache.geode.internal.cache.execute.MemberFunctionExecutor.executeFunction(MemberFunctionExecutor.java:191)
        at org.apache.geode.internal.cache.execute.AbstractExecution.execute(AbstractExecution.java:397)
        at org.apache.geode.internal.cache.execute.AbstractExecution.execute(AbstractExecution.java:402)
        at org.apache.geode.modules.util.BootstrappingFunction.bootstrapMember(BootstrappingFunction.java:170)
        at org.apache.geode.modules.util.BootstrappingFunction.memberJoined(BootstrappingFunction.java:240)
        at org.apache.geode.distributed.internal.ClusterDistributionManager$MemberJoinedEvent.handleEvent(ClusterDistributionManager.java:2497)
        at org.apache.geode.distributed.internal.ClusterDistributionManager$MemberEvent.handleEvent(ClusterDistributionManager.java:2450)
        at org.apache.geode.distributed.internal.ClusterDistributionManager$MemberEvent.handleEvent(ClusterDistributionManager.java:2439)
        at org.apache.geode.distributed.internal.ClusterDistributionManager.handleMemberEvent(ClusterDistributionManager.java:1405)
        at org.apache.geode.distributed.internal.ClusterDistributionManager.access$200(ClusterDistributionManager.java:109)
        at org.apache.geode.distributed.internal.ClusterDistributionManager$MemberEventInvoker.run(ClusterDistributionManager.java:1437)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

We do check if a shutdown is in progress before getting to this if block and throw an exception there, so that case would not end up logging this message:
```
if (membership.shutdownInProgress()) {
        checkCancelled();
        throw new DistributedSystemDisconnectedException();
      }
```

We also catch `MembershipClosedException` and `DistributedSystemDisconnectedException` and throw exceptions for those (so they'd never log this message).